File: ruby/y2025/day09.rb
#!/usr/bin/env ruby
#
# Movie Theater

require_relative '../day'
require_relative '../point'

class Day09 < Day
  def do_part1(lines)
    parse_2d(lines)
      .combination(2)
      .map { |p0, p1| p2_area(p0, p1) }
      .max
  end

  # A rectangle is an array of [x_min, x_max, y_min, y_max].
  # An edge is an array of [p0_x, p0_y, p1_x, p1_y].
  def do_part2(lines)
    red_tiles = parse_2d(lines)
    # edges = red_tiles.each_cons(2).to_a + [[red_tiles[-1], red_tiles[0]]]
    # horiz_edges, vert_edges = edges.partition { |edge| edge[0].y == edge[1].y }

    edges = (red_tiles.each_cons(2).to_a + [[red_tiles[-1], red_tiles[0]]])
            .map do |p0, p1|
      x_min, x_max = [p0, p1].map(&:x).minmax
      y_min, y_max = [p0, p1].map(&:y).minmax
      [x_min, y_min, x_max, y_max]
    end

    max_area = 0
    red_tiles
      .combination(2)                     # two-corner pairs of rect corners
      .map { |p0, p1| rect_from(p0, p1) } # produce array of [x_min, y_min, x_max, y_max]
      # .select { |rect| rect == [2, 1, 11, 5] }
      .each do |rect| # calc area, if > max and in-bounds save area else reject
        area = rect_area(rect)
        max_area = area if (area > max_area) && !crosses_any_edge?(rect, edges)
      end
    max_area
  end

  private

  # Returns an array of [x_min, y_min, x_max, y_max].
  def rect_from(p0, p1)
    x_min, x_max = [p0, p1].map(&:x).minmax
    y_min, y_max = [p0, p1].map(&:y).minmax
    [x_min, y_min, x_max, y_max]
  end

  def crosses_any_edge?(rect, edges)
    x_min, y_min, x_max, y_max = *rect

    edges.any? do |edge|
      e_x_min, e_x_max = [edge[0], edge[2]].minmax
      e_y_min, e_y_max = [edge[1], edge[3]].minmax
      x_min < e_x_max && x_max > e_x_min && y_min < e_y_max && y_max > e_y_min
    end
  end

  # Given two corners of a rectangle, return the area
  def p2_area(p0, p1)
    ((p1.x - p0.x).abs + 1) * ((p1.y - p0.y).abs + 1)
  end

  # Given two corners of a rectangle, return the area
  def rect_area(rect)
    x_min, y_min, x_max, y_max = *rect
    (x_max - x_min + 1) * (y_max - y_min + 1)
  end

  def parse_2d(lines)
    lines.map { |line| Point.new(*line.split(',').map(&:to_i)) }
  end
end

if __FILE__ == $PROGRAM_NAME
  require_relative '../aoc'

  aoc(__FILE__)
end

File: elixir/lib/y2025/day09.ex
# ================ Movie Theater ================

defmodule Y2025.Day09 do
  alias Common.Enum, as: CE

  def part1(_ctx, lines) do
    lines
    |> Enum.map(&parse_2d/1)
    |> CE.combinations(2)
    |> Enum.map(fn [p0, p1] -> p2_area(p0, p1) end)
    |> Enum.max()
  end

  # A rectangle is a min/max tuple of {x_min, y_min, x_max, y_max}.
  # An edge is a tuple of two points {{x0, y0}, {x1, y1}}.
  def part2(_ctx, lines) do
    red_tiles = lines |> Enum.map(&parse_2d/1)

    edges =
      [List.last(red_tiles) | red_tiles]
      |> CE.window(2)
      |> Enum.map(fn [{x0, y0}, {x1, y1}] ->
        {x_min, x_max} = Enum.min_max([x0, x1])
        {y_min, y_max} = Enum.min_max([y0, y1])
        {{x_min, y_min}, {x_max, y_max}}
      end)

    red_tiles
    |> CE.combinations(2)
    |> Enum.map(fn [p1, p2] -> rect_from(p1, p2) end)
    |> Enum.reduce(0, fn rect, max_area ->
      area = rect_area(rect)
      if area > max_area and not crosses_any_edge?(rect, edges), do: area, else: max_area
    end)
  end

  # ================ helpers ================

  # Returns the area of the rect under two corners.
  defp p2_area({x0, y0}, {x1, y1}) do
    (abs(x1 - x0) + 1) * (abs(y1 - y0) + 1)
  end

  # Given two corners of a rectangle, return the area.
  defp rect_area({x_min, y_min, x_max, y_max}) do
    (x_max - x_min + 1) * (y_max - y_min + 1)
  end

  # Returns a tuple of {x_min, y_min, x_max, y_max}.
  defp rect_from({x0, y0}, {x1, y1}) do
    [x_min, x_max] = Enum.sort([x0, x1])
    [y_min, y_max] = Enum.sort([y0, y1])
    {x_min, y_min, x_max, y_max}
  end

  defp crosses_any_edge?({x_min, y_min, x_max, y_max}, edges) do
    edges
    |> Enum.any?(fn {{e_x0, e_y0}, {e_x1, e_y1}} ->
      {e_x_min, e_x_max} = Enum.min_max([e_x0, e_x1])
      {e_y_min, e_y_max} = Enum.min_max([e_y0, e_y1])
      x_min < e_x_max && x_max > e_x_min && y_min < e_y_max && y_max > e_y_min
    end)
  end

  # Returns an {x, y} tuple.
  defp parse_2d(line) do
    line
    |> String.split(",")
    |> Enum.map(&String.to_integer/1)
    |> List.to_tuple()
  end
end
